programs/compressed-pda/tests/test.rs

```diff
@@ -478,17 +451,18 @@ async fn test_with_address() {
     )
     .await;
     // Should fail to insert the same address twice in the same tx
-    assert_eq!(
-        res.result,
-        Err(solana_sdk::transaction::TransactionError::InstructionError(
-            0,
-            InstructionError::Custom(light_hash_set::HashSetError::ElementAlreadyExists.into())
+    println!("EVENT ERR: {event:?}");
+    assert!(matches!(
+        event,
+        Err(BanksClientError::TransactionError(
+            // ElementAlreadyExists
+            TransactionError::InstructionError(0, InstructionError::Custom(6002))
```

Comment:
this is weird the error shouldn't change

Reply:
The error didn't change. The way how I'm asserting it changed. Unfortunately, I had to generalize the error returned in `create_and_send_transaction_with_event` to `BanksClientError` - which has `TransactionError` as a variant. `BanksClientError` covers all the error types which are thrown in that function and not all of them are of `TransactionError` variant.

Such a generalization comes with a downside - while [`TransactionError`](https://docs.rs/solana-sdk/latest/solana_sdk/transaction/enum.TransactionError.html) implements `Eq`, [`BanksError`](https://docs.rs/solana-banks-client/latest/solana_banks_client/enum.BanksClientError.html) doesn't. So I cannot use `assert_eq` on `BanksClientError`. Attempt to do so ends up with the following error:

```
error[E0369]: binary operation `==` cannot be applied to type `Result<Option<PublicTransactionEvent>, BanksClientError>`
   --> programs/compressed-pda/tests/test.rs:452:5
    |
452 | /     assert_eq!(
453 | |         event,
454 | |         Err(BanksClientError::TransactionError(
455 | |             TransactionError::InstructionError(0, InstructionError::Custom(HashSetError::ElementAlreadyExists.into()))
456 | |         ))
457 | |     );
    | |     ^
    | |_____|
    | |_____Result<Option<PublicTransactionEvent>, BanksClientError>
    |       Result<Option<PublicTransactionEvent>, BanksClientError>
    |
    = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Usually the way to assert error types which don't implement `Eq` is to use `matches!`. But `matches!` comes with its own limitations - like, not allowing non-exhaustive function calls inside the pattern.

Therefore, trying to use `HashSetError::ElementAlreadyExists.into()` inside `matches!` throws:

```
error: expected one of `!`, `(`, `)`, `,`, `...`, `..=`, `..`, `::`, `{`, or `|`, found `.`
   --> programs/compressed-pda/tests/test.rs:467:76
    |
467 |                 InstructionError::Custom(HashSetError::ElementAlreadyExists.into())
    |                                                                            ^
    |                                                                            |
    |                                                                            expected one of 10 possible tokens
    |                                                                            help: missing `,`

error[E0531]: cannot find tuple struct or tuple variant `into` in this scope
   --> programs/compressed-pda/tests/test.rs:467:77
    |
467 |                 InstructionError::Custom(HashSetError::ElementAlreadyExists.into())
    |                                                                             ^^^^ not found in this scope

error[E0023]: this pattern has 2 fields, but the corresponding tuple variant has 1 field
   --> programs/compressed-pda/tests/test.rs:467:42
    |
467 |                 InstructionError::Custom(HashSetError::ElementAlreadyExists.into())
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^ expected 1 field, found 2
    |
   ::: /home/node/.cargo/git/checkouts/agave-f8c52564d46c9282/a24d3c0/sdk/program/src/instruction.rs:142:12
    |
142 |     Custom(u32),
    |            --- tuple variant has 1 field
```

Pattern match expects me to put just an integer constant there. Which is what I did here.

I will try to quickly think of alternatives, but so far this is the least ugly way I came up with.

Reply:
tried a couple things that all didn't work, cannot think of another way either